### PR TITLE
Speedup environment image and jupyter builds by using `buildah` for image builds

### PR DESCRIPTION
**Pull request by fruttasecca**
## Description

This PR replaces `buildkit` in favour of `buildah` for image building to speedup said builds, also refactors the way we do logging for image builds and removes any mention of `kaniko` and `buildkit` from the codebase in b4fb56ea. Most of the speed up comes from being able to avoid compression when pushing layers to the local registry.

The on-disk image builder caching now also covers the pip cache, which is mounted during the build, making it so that said cache is persisted across builds; this allows repeating builds without continuously pulling packages. I think we want to do the same for other package managers if possible.

Example of logs for a setup script containing `pip install pandas scikit-learn matplotlib --quiet`:
```bash
Starting worker...
Starting image build.............
Copying context....
Running environment set-up script...
Storing image................
Done!
```

### CODE REVIEW

File: services/orchest-api/app/app/core/image_utils.py [hunk 1]
```diff
@@ -221,254 +163,173 @@ def _get_buildkit_image_build_workflow_manifest(
...
-    all_logs.flush()
+    The class is implemented as a state machine. The only real "state"
+    it has are the log files to which to write logs and the current
+    state, reflected by self._log_handler_function. Rough schema:
 
+    Starting──►Building───►Copying──┐
+    Worker     Image  │    Context  └►Running──────►Storing
+                      ▼    ▲          Setup script  Image
+                    Pulling│
+                    Base Image
```
Comment (yannickperrenet) [inline services/orchest-api/app/app/core/image_utils.py:177]: Is this a plantUML export? Or how did you create it? Definitely great to be able to include these images in the source code directly (just need to come up with a good way to maintain them as well).

Comment (fruttasecca) [reply services/orchest-api/app/app/core/image_utils.py:177]: Just some quick and dirty drawing using https://asciiflow.com/#/

File: services/orchest-api/app/app/core/image_utils.py [hunk 2]
```diff
@@ -124,37 +64,34 @@ def _get_buildkit_image_build_workflow_manifest(
...
+                                f"--tag {full_image_name} "
+                                # Push
+                                "&& buildah push "
+                                "--disable-compression=true "
+                                # Buildah might compress regardless of
+                                # the specified options depending on the
+                                # destination storage, tune such
+                                # compression.
+                                "--compression-format=zstd:chunked "
+                                "--compression-level=0 "
```
[the remark this answers is not in the document]
Comment (fruttasecca) [reply services/orchest-api/app/app/core/image_utils.py:92]: I haven't tried to be honest, I kept those flags for compatibility for when we move back to officially released images, will test tomorrow.

File: services/orchest-api/app/app/core/image_utils.py [hunk 3]
```diff
@@ -168,7 +105,13 @@ def _get_buildkit_image_build_workflow_manifest(
                             },
                             {
                                 "name": "image-builder-cache-pvc",
-                                "mountPath": "/cache",
+                                "subPath": "pip-cache",
+                                "mountPath": "/pip-cache",
```
[the remark this answers is not in the document]
Comment (fruttasecca) [reply services/orchest-api/app/app/core/image_utils.py:109]: Indeed I'll try to see if we can do the same for mamba

**Comment by yannickperrenet** [comment]
This comment is now outdated https://github.com/orchest/orchest/blob/80a883380fc7fa262b0658f81642558952390287/services/orchest-api/app/app/apis/namespace_ctl.py#L65

File: services/orchest-api/app/app/core/image_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fruttasecca) [reply services/orchest-api/app/app/core/image_utils.py:92]: Removing those flags doesn't seem to break anything but I would still rather keep em for the aforementioned reasons, or at least `--disable-compression`.

**Comment by fruttasecca** [comment]
>How about changing the "Building..." depending on the step in the buildprocess, e.g. "Storing image..."? (Not for this PR, but to open an issue.)

That would be a nice to have but imo a low ROI right now

**Comment by fruttasecca** [comment]
>Regarding installing with mamba, there was no "feedback" that the install was actually working as no new log lines were added for a significant time.

I think that might have been because of mamba, see video.

https://www.loom.com/share/f51f015762bd4b59b56a298143ff9738

File: services/orchest-api/app/app/core/image_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fruttasecca) [reply services/orchest-api/app/app/core/image_utils.py:114]: `mamba`'s (and `conda`, they share the same cache) cache is now persisted across builds

**Comment by fruttasecca** [comment]
>Since the pip-cache is shared among all projects/environments, couldn't there be certain issues when installing packages? For example when one environment requires a specific version? Not sure how pip resolves dependencies if it finds some in cache (will it still download the correct version as required?).

I don't think this is an issue given that the cache is pretty much just storage AFAIK, nevertheless I am looking for edge cases. Small example
```bash
pip install pandas==1.4.2
pip install pandas==1.4.1
```

```
Running environment set-up script...
test
Collecting pandas==1.4.2
  Using cached pandas-1.4.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.7 MB)
Requirement already satisfied: numpy>=1.18.5 in /opt/conda/lib/python3.9/site-packages (from pandas==1.4.2) (1.22.3)
Requirement already satisfied: python-dateutil>=2.8.1 in /opt/conda/lib/python3.9/site-packages (from pandas==1.4.2) (2.8.2)
Requirement already satisfied: pytz>=2020.1 in /opt/conda/lib/python3.9/site-packages (from pandas==1.4.2) (2021.3)
Requirement already satisfied: six>=1.5 in /opt/conda/lib/python3.9/site-packages (from python-dateutil>=2.8.1->pandas==1.4.2) (1.16.0)
Installing collected packages: pandas
Successfully installed pandas-1.4.2
Collecting pandas==1.4.1
  Using cached pandas-1.4.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.7 MB)
Requirement already satisfied: numpy>=1.18.5 in /opt/conda/lib/python3.9/site-packages (from pandas==1.4.1) (1.22.3)
Requirement already satisfied: pytz>=2020.1 in /opt/conda/lib/python3.9/site-packages (from pandas==1.4.1) (2021.3)
Requirement already satisfied: python-dateutil>=2.8.1 in /opt/conda/lib/python3.9/site-packages (from pandas==1.4.1) (2.8.2)
Requirement already satisfied: six>=1.5 in /opt/conda/lib/python3.9/site-packages (from python-dateutil>=2.8.1->pandas==1.4.1) (1.16.0)
Installing collected packages: pandas
  Attempting uninstall: pandas
    Found existing installation: pandas 1.4.2
    Uninstalling pandas-1.4.2:
      Successfully uninstalled pandas-1.4.2
Successfully installed pandas-1.4.1
Storing image.........
Done!
```

**Comment by yannickperrenet** [comment]
> nevertheless I am looking for edge cases.

One thing I could come up with is:

project A:
```
pip install pandas
```

project B:
```
pip install pandas==1.2.0
```

Actions taken:
1. Project A builds the environment and gets latest pandas version
2. Project B builds the environment changing the pandas version in the cache
3. Project A builds and uses the pandas version from the cache instead of the latest pandas version. Now user code that uses latest pandas features will suddenly break due to what happened in the environment build in another project.

**Comment by fruttasecca** [comment]
>Project B builds the environment changing the pandas version in the cache

About this point, both versions will be in the cache. I have just tried this out btw, and the second round of `pip install pandas` installed the latest pandas version that was in the cache.

**Comment by ricklamers** [comment]
Nicely done @fruttasecca! LGTM
